Storage/tests/Unit/Connection/RestTest.php

```diff
@@ -469,6 +471,237 @@ public function validationMethod()
         ];
     }
 
+    /**
+     * @dataProvider retryFunctionReturnValues
+     */
+    public function testIsPreconditionSupplied(
+        $resource,
+        $op,
+        $restConfig,
+        $args,
+        $errorCode,
+        $currAttempt,
+        $expected
+    ) {
+        // Using Reflection instead of Prophecy because we want to test a
+        // private method's logic by verifying the output for a given input.
+        $rest = new Rest($restConfig);
+        $reflector = new \ReflectionClass('Google\Cloud\Storage\Connection\Rest');
+        $method = $reflector->getMethod('isPreConditionSupplied');
+        $condIdempotentOps = $reflector->getProperty('condIdempotentOps');
+        $method->setAccessible(true);
+        $condIdempotentOps->setAccessible(true);
+        $condIdempotentOps = $condIdempotentOps->getValue($rest);
+        $methodName = sprintf('%s.%s', $resource, $op);
+        $preconditionNeeded = array_key_exists($methodName, $condIdempotentOps);
+        $result = $method->invokeArgs($rest, array($methodName, $args));
+        if ($preconditionNeeded and $errorCode != 400) {
+            $this->assertEquals($result, $expected);
```

Comment:
The tests are incredibly hard to follow. I've spent a while trying to decipher the data provider, which is built off another data provider, and then there's still conditionals in that data provider and also in the test-case itself. 

These need to be broken apart into individual tests which make sense on their own (and test one specific thing). What we have here is something that seems to test everything at once, and it's impossible to follow.

Reply:
Thanks for the callout.

In the latest commit, I have split the tests of the retry function and the retry strategy.

They should be easy to read now.